runtime/warp.cu

```diff
@@ -40,19 +59,26 @@ __device__ void warpReduceTIDX(
   }
 
   // Reduce within each warp
+  // Register usage is reduced when Padded is true due to the elimination of the
+  // if-statement.
+  unsigned int warp_idx = threadIdx.x / WARP_SIZE;
+  unsigned int lane_idx = threadIdx.x % WARP_SIZE;
+  unsigned int reduction_size = blockDim.x;
+  unsigned int num_of_warps = (reduction_size + WARP_SIZE - 1) / WARP_SIZE;
+  int hoist_idx = reduction_size - lane_idx - warp_idx * WARP_SIZE;
   for (int i = 16; i >= 1; i /= 2) {
-    reduction_op(reduce_val, shfl_xor(reduce_val, i, WARP_SIZE));
+    T shuffled_value = shfl_xor(reduce_val, i, WARP_SIZE);
+    if (Padded || i < hoist_idx) {
+      reduction_op(reduce_val, shuffled_value);
+    }
```

Comment:
Hmm, does this really work when, for example, `blockDim.x == 16, blockDim.y == 2`? Wouldn't it result in reducing across all threads within a thread block?

Also, is there any guarantee that the total number of threads per block is a multiple of warps? Don't we need that to use the warp instructions?

Reply:
(1) `blockDim.x == 16, blockDim.y == 2` is not supposed to use this warp reduction. The unpadded version only works if the Tv is not `isParallelizedByTidyOrTidz`, that is `blockDim.y == blockDim.z == 1`

(2) There is no guarantee that the total number of threads per block is a multiple of warps. That's why I need the if-statement to avoid using undefined data in other threads. The hardware will patch to a warp and the reduction code ensures the data in these patched threads is not being used.

Reply:
>(1) blockDim.x == 16, blockDim.y == 2 is not supposed to use this warp reduction. The unpadded version only works if the Tv is not isParallelizedByTidyOrTidz, that is blockDim.y == blockDim.z == 1

Then, do not just remove the comment about the assumption. Make sure to document when this function should be used under what assumptions.